File: src/cards.js
// src/cards.js
// 「飙分」牌模型:108 张(两副),主花色分类与组内大小排序。
// 详细规则见 飙分-规则.md。

export const SUITS = ['S', 'H', 'D', 'C'];                 // ♠ ♥ ♦ ♣
export const SUIT_SYMBOL = { S: '♠', H: '♥', D: '♦', C: '♣', JOKER: '🃏' };

export const SMALL_JOKER = 16;
export const BIG_JOKER = 17;

export const RANK_LABEL = {
  2: '2', 3: '3', 4: '4', 5: '5', 6: '6', 7: '7', 8: '8', 9: '9', 10: '10',
  11: 'J', 12: 'Q', 13: 'K', 14: 'A', [SMALL_JOKER]: '小王', [BIG_JOKER]: '大王',
};

export function makeCard(suit, rank, copy = 0) {
  return { suit, rank, copy, id: `${suit}-${rank}-${copy}` };
}

export function isJoker(card) { return card.suit === 'JOKER'; }

export function cardLabel(card) {
  return isJoker(card) ? RANK_LABEL[card.rank] : SUIT_SYMBOL[card.suit] + RANK_LABEL[card.rank];
}

// 一副 54 张(52 + 大小王),两副共 108
export function buildDeck() {
  const cards = [];
  for (let copy = 0; copy < 2; copy++) {
    for (const suit of SUITS)
      for (let rank = 2; rank <= 14; rank++)
        cards.push(makeCard(suit, rank, copy));
    cards.push(makeCard('JOKER', SMALL_JOKER, copy));
    cards.push(makeCard('JOKER', BIG_JOKER, copy));
  }
  return cards;
}

// 常主:大小王、所有 3、所有 2;再加主花色全部 → 主牌
export function isTrump(card, trumpSuit) {
  if (isJoker(card)) return true;
  if (card.rank === 3 || card.rank === 2) return true;
  return card.suit === trumpSuit;
}

// 分组:'TRUMP' 或 边花色字母('H'/'D'/'C'/'S')
export function cardGroup(card, trumpSuit) {
  return isTrump(card, trumpSuit) ? 'TRUMP' : card.suit;
}

// 组内「比大小」刻度(越大越强)——决定谁压谁。与「拖拉机相邻」是两套规则(见 tractorRank)。
// 主牌:大王200 > 小王199 > 主3 198 > 副3 197 > 主2 196 > 副2 195 > 主A 194 > 主K 193 …… 主4 184。
// 即所有 3 > 所有 2;3 内主3>副3、2 内主2>副2(主花色那张更大)。边牌按点数 4..14。
export function strength(card, trumpSuit) {
  if (cardGroup(card, trumpSuit) !== 'TRUMP') return card.rank;   // 边牌:4..14
  if (card.rank === BIG_JOKER) return 200;
  if (card.rank === SMALL_JOKER) return 199;
  const mainSuit = card.suit === trumpSuit;
  if (card.rank === 3) return mainSuit ? 198 : 197;   // 主3 > 副3
  if (card.rank === 2) return mainSuit ? 196 : 195;   // 主2 > 副2(且所有3 > 所有2)
  return card.rank + 180;            // 主花色 4..14(A) → 184..194(均 < 副2 195)
}

// 「拖拉机相邻」刻度——只管连不连,与比大小分开。返回 { lane, pos, alt? }:
//   同 lane(同花色/同孤岛)且位置(pos)差 1 才相连。
//   主花色:ace-low 自然阶梯 A=1,2=2,3=3,…,K=13;主 A「两头都连」→ pos:1 兼 alt:14(接 2 也接 K)。
//   王:小王/大王 自成孤岛(lane 'JOKER',pos 16/17 相邻)。边牌:ace-high,pos = 点数 4..14。
export function tractorRank(card, trumpSuit) {
  if (isJoker(card)) return { lane: 'JOKER', pos: card.rank };          // 16、17 相邻
  if (isTrump(card, trumpSuit)) {
    if (card.rank === 14) return { lane: 'T-' + card.suit, pos: 1, alt: 14 }; // 主A:接2 或 接K(A 必为主花色)
    return { lane: 'T-' + card.suit, pos: card.rank };                 // 主3=3、主2=2、…、主K=13
  }
  return { lane: card.suit, pos: card.rank };                          // 边牌 ace-high 4..14
}

export function shuffle(cards, rng = Math.random) {
  const a = cards.slice();
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}

// 发牌。players: 3 或 4。返回 { hands, kitty, kittySize, perHand }
// 4 人 → 每人 25、底牌 8;3 人 → 每人 33、底牌 9。
export function deal(cards, players) {
  const kittySize = players === 4 ? 8 : 9;
  const perHand = (cards.length - kittySize) / players;
  const hands = Array.from({ length: players }, () => []);
  let i = 0;
  for (let p = 0; p < players; p++)
    for (let k = 0; k < perHand; k++) hands[p].push(cards[i++]);
  return { hands, kitty: cards.slice(i), kittySize, perHand };
}

// 整理手牌(展示用):主牌在前,组内从大到小
export function sortHand(hand, trumpSuit) {
  const groupOrder = { TRUMP: 0, S: 1, H: 2, D: 3, C: 4 };
  const suitOrder = { S: 0, H: 1, D: 2, C: 3, JOKER: 4 };
  return hand.slice().sort((a, b) => {
    const ga = cardGroup(a, trumpSuit), gb = cardGroup(b, trumpSuit);
    if (groupOrder[ga] !== groupOrder[gb]) return groupOrder[ga] - groupOrder[gb];
    const sd = strength(b, trumpSuit) - strength(a, trumpSuit);
    if (sd !== 0) return sd;
    // 同强度(如不同花色的 3)→ 按花色、点数、copy 归拢,让相同的牌永远相邻(对子不被拆开)
    if (a.suit !== b.suit) return suitOrder[a.suit] - suitOrder[b.suit];
    if (a.rank !== b.rank) return b.rank - a.rank;
    return a.copy - b.copy;
  });
}

File: server/game/cards.go
// Package game — 「飙分」规则引擎(1:1 移植自 src/cards.js 等,规则见 飙分-规则.md)。
package game

import (
	"fmt"
	"math/rand"
	"sort"
)

const (
	SmallJoker = 16
	BigJoker   = 17
)

var Suits = []string{"S", "H", "D", "C"} // ♠ ♥ ♦ ♣

// Card 的 JSON 字段与前端 src/cards.js 完全一致(id 格式 "S-5-0"),前后端直接互通。
type Card struct {
	Suit string `json:"suit"`
	Rank int    `json:"rank"`
	Copy int    `json:"copy"`
	ID   string `json:"id"`
}

func MakeCard(suit string, rank, copy int) Card {
	return Card{Suit: suit, Rank: rank, Copy: copy, ID: fmt.Sprintf("%s-%d-%d", suit, rank, copy)}
}

func (c Card) IsJoker() bool { return c.Suit == "JOKER" }

// 一副 54 张(52 + 大小王),两副共 108
func BuildDeck() []Card {
	cards := make([]Card, 0, 108)
	for copy := 0; copy < 2; copy++ {
		for _, s := range Suits {
			for r := 2; r <= 14; r++ {
				cards = append(cards, MakeCard(s, r, copy))
			}
		}
		cards = append(cards, MakeCard("JOKER", SmallJoker, copy))
		cards = append(cards, MakeCard("JOKER", BigJoker, copy))
	}
	return cards
}

// 常主:大小王、所有 3、所有 2;再加主花色全部 → 主牌
func IsTrump(c Card, trumpSuit string) bool {
	if c.IsJoker() {
		return true
	}
	if c.Rank == 3 || c.Rank == 2 {
		return true
	}
	return c.Suit == trumpSuit
}

// 分组:"TRUMP" 或 边花色字母("H"/"D"/"C"/"S")
func CardGroup(c Card, trumpSuit string) string {
	if IsTrump(c, trumpSuit) {
		return "TRUMP"
	}
	return c.Suit
}

// 组内「比大小」刻度(越大越强)——决定谁压谁。与「拖拉机相邻」是两套规则(见 TractorRankOf)。
// 主牌:大王200 > 小王199 > 主3 198 > 副3 197 > 主2 196 > 副2 195 > 主A 194 …… 主4 184。
func Strength(c Card, trumpSuit string) int {
	if CardGroup(c, trumpSuit) != "TRUMP" {
		return c.Rank // 边牌:4..14
	}
	if c.Rank == BigJoker {
		return 200
	}
	if c.Rank == SmallJoker {
		return 199
	}
	mainSuit := c.Suit == trumpSuit
	switch c.Rank {
	case 3:
		if mainSuit {
			return 198
		}
		return 197 // 主3 > 副3
	case 2:
		if mainSuit {
			return 196
		}
		return 195 // 主2 > 副2(且所有3 > 所有2)
	}
	return c.Rank + 180 // 主花色 4..14(A) → 184..194(均 < 副2 195)
}

// TRank 「拖拉机相邻」刻度——只管连不连,与比大小分开。
// 同 Lane(同花色/同孤岛)且 Pos 差 1 才相连。Alt=0 表示无备选位;
// 主 A「两头都连」→ Pos:1 兼 Alt:14(接 2 也接 K)。王自成孤岛(16/17 相邻)。
type TRank struct {
	Lane string
	Pos  int
	Alt  int
}

func TractorRankOf(c Card, trumpSuit string) TRank {
	if c.IsJoker() {
		return TRank{Lane: "JOKER", Pos: c.Rank}
	}
	if IsTrump(c, trumpSuit) {
		if c.Rank == 14 {
			return TRank{Lane: "T-" + c.Suit, Pos: 1, Alt: 14} // 主A(A 必为主花色)
		}
		return TRank{Lane: "T-" + c.Suit, Pos: c.Rank} // 主3=3、主2=2、…、主K=13
	}
	return TRank{Lane: c.Suit, Pos: c.Rank} // 边牌 ace-high 4..14
}

func Shuffle(cards []Card, rng *rand.Rand) []Card {
	a := append([]Card(nil), cards...)
	for i := len(a) - 1; i > 0; i-- {
		j := rng.Intn(i + 1)
		a[i], a[j] = a[j], a[i]
	}
	return a
}

type DealResult struct {
	Hands     [][]Card
	Kitty     []Card
	KittySize int
	PerHand   int
}

// 发牌。players: 3 或 4。4 人 → 每人 25、底牌 8;3 人 → 每人 33、底牌 9。
func Deal(cards []Card, players int) DealResult {
	kittySize := 9
	if players == 4 {
		kittySize = 8
	}
	perHand := (len(cards) - kittySize) / players
	hands := make([][]Card, players)
	i := 0
	for p := 0; p < players; p++ {
		hands[p] = make([]Card, 0, perHand)
		for k := 0; k < perHand; k++ {
			hands[p] = append(hands[p], cards[i])
			i++
		}
	}
	return DealResult{Hands: hands, Kitty: append([]Card(nil), cards[i:]...), KittySize: kittySize, PerHand: perHand}
}

// SideSuitOrder 边花色展示顺序:按主花色动态排列,保证相邻花色黑红交错(不易看错)
func SideSuitOrder(trumpSuit string) []string {
	switch trumpSuit {
	case "S":
		return []string{"H", "C", "D"} // 红黑红
	case "C":
		return []string{"H", "S", "D"} // 红黑红
	case "H":
		return []string{"S", "D", "C"} // 黑红黑
	case "D":
		return []string{"S", "H", "C"} // 黑红黑
	}
	return []string{"S", "H", "C", "D"} // 未亮主:黑红黑红
}

// 整理手牌(展示用):主牌在前,组内从大到小;相同的牌相邻(对子不被拆)
func SortHand(hand []Card, trumpSuit string) []Card {
	groupOrder := map[string]int{"TRUMP": 0}
	for i, s := range SideSuitOrder(trumpSuit) {
		groupOrder[s] = i + 1
	}
	suitOrder := map[string]int{"S": 0, "H": 1, "D": 2, "C": 3, "JOKER": 4}
	a := append([]Card(nil), hand...)
	sort.SliceStable(a, func(i, j int) bool {
		ga, gb := CardGroup(a[i], trumpSuit), CardGroup(a[j], trumpSuit)
		if groupOrder[ga] != groupOrder[gb] {
			return groupOrder[ga] < groupOrder[gb]
		}
		sa, sb := Strength(a[i], trumpSuit), Strength(a[j], trumpSuit)
		if sa != sb {
			return sa > sb
		}
		if a[i].Suit != a[j].Suit {
			return suitOrder[a[i].Suit] < suitOrder[a[j].Suit]
		}
		if a[i].Rank != a[j].Rank {
			return a[i].Rank > a[j].Rank
		}
		return a[i].Copy < a[j].Copy
	})
	return a
}
